File: Day35/Part2/kth_smallest_pair_distance.cpp
#include <bits/stdc++.h>
using namespace std;

long long count_pairs(const vector<int>& a, int d){
    long long cnt=0; int n=(int)a.size(); int j=0;
    for(int i=0;i<n;++i){
        while(j<n && a[j]-a[i]<=d) ++j;
        cnt += max(0, j - i - 1);
    }
    return cnt;
}

int smallest_distance_pair(vector<int> nums, int k){
    sort(nums.begin(), nums.end());
    int lo=0, hi=nums.back()-nums.front();
    while(lo<hi){
        int mid = lo + (hi-lo)/2;
        if(count_pairs(nums, mid) >= k) hi=mid; else lo=mid+1;
    }
    return lo;
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout << smallest_distance_pair(vector<int>{1,3,1}, 1) << "\n"; // Expected 0
    return 0;
}

File: Day35/Part2/kth_smallest_pair_distance.py
from typing import List

def count_pairs(nums: List[int], d: int) -> int:
    n = len(nums)
    j = 0
    cnt = 0
    for i in range(n):
        while j < n and nums[j] - nums[i] <= d:
            j += 1
        cnt += max(0, j - i - 1)
    return cnt

def smallest_distance_pair(nums: List[int], k: int) -> int:
    nums.sort()
    lo, hi = 0, nums[-1] - nums[0]
    while lo < hi:
        mid = (lo + hi) // 2
        if count_pairs(nums, mid) >= k:
            hi = mid
        else:
            lo = mid + 1
    return lo

if __name__ == "__main__":
    print(smallest_distance_pair([1,3,1], 1))  # Expected 0
